### PR TITLE
Increase the size of the API and add an actuall API.

### PR DESCRIPTION
**Pull request by marchermans**
This PR adds a lot of interfaces and connections in between them to allow for a larger API surface.
Beware this API is not complete, a lot of functionality is missing, and some interfaces should possibly be accompanied by a set of abstract classes but haven't, so keep that in mind while reviewing.

Not much logic has changed, with the exception of some small things like the `JobRegistry`. Which shows an example of how to prevent things like reflection and other not so nice things which should not be used in an api.

Also this is a split of my original PR #3747 which just the API rework, part of why i made this, is so that we can write better and more unit tests, which should allow less mistakes and mishaps to pass.

People who need to review this, have a review requested. Those are in particular people who have made PRs in the last couple of weeks.
People who have not the rights to review, but should take a look:
@Elrol 
@Mystical-Codez 
@phit

### CODE REVIEW

**Comment by marchermans** [comment]
Funny i signed the CLA a long time ago, and it will not allow me to resign it again......

**Comment by Raycoms** [comment]
I think it looks very good.
There are some things we'll have to work on top of it, but we can do this in separate PRs (pulling javadoc to all interfaces for example).

I'd say :shipit: but build is failing and codacy is not happy at all.

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/3756/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

File: src/api/java/com/minecolonies/coremod/entity/IBaseEntityCitizen.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraftforge.common.capabilities.ICapabilitySerializable;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Random;
+import java.util.UUID;
+
+public interface IBaseEntityCitizen extends ICommandSender, ICapabilitySerializable<NBTTagCompound>, INpc
```
Comment (someaddons) [inline src/api/java/com/minecolonies/coremod/entity/IBaseEntityCitizen.java:27]: whatfor is the ICommandSender needed? so they can add chat messages?

Comment (marchermans) [reply src/api/java/com/minecolonies/coremod/entity/IBaseEntityCitizen.java:27]: I guess so, this was on the original class so I just pulled it up into the interface.

That said I am not entirely happy with the api on the citizen level yet, but that is for another pr

**Review by Asherslab — APPROVED** [review]
From my readthrough i don't see any issues =D

**Comment by Asherslab** [comment]
This PR was reverted. Caused crashes.
Details here: https://github.com/ldtteam/minecolonies/pull/3759

**Pull request merged**